src/DataGrid.tsx

```diff
@@ -617,29 +597,44 @@ function DataGrid<R, SR, K extends Key>(
     updateRow(columns[selectedPosition.idx], selectedPosition.rowIdx, selectedPosition.row);
   }
 
-  function handleCopy() {
+  function handleCopy(event: React.ClipboardEvent<HTMLDivElement>) {
+    if ((!onPaste && !onCopy) || !selectedCellIsWithinViewportBounds) return;
     const { idx, rowIdx } = selectedPosition;
     const sourceRow = rows[rowIdx];
     const sourceColumnKey = columns[idx].key;
+    const cellEvent = createCellEvent(event);
+    onCopy?.({ sourceRow, sourceColumnKey }, cellEvent);
+    if (cellEvent.isGridDefaultPrevented()) return;
+
     setCopiedCell({ row: sourceRow, columnKey: sourceColumnKey });
-    onCopy?.({ sourceRow, sourceColumnKey });
   }
 
-  function handlePaste() {
-    if (!onPaste || !onRowsChange || copiedCell === null || !isCellEditable(selectedPosition)) {
+  function handlePaste(event: React.ClipboardEvent<HTMLDivElement>) {
+    if (
+      !onPaste ||
+      !onRowsChange ||
+      copiedCell === null ||
+      !isCellEditable(selectedPosition) ||
+      !selectedCellIsWithinViewportBounds
+    ) {
       return;
     }
 
     const { idx, rowIdx } = selectedPosition;
     const targetColumn = columns[idx];
     const targetRow = rows[rowIdx];
 
-    const updatedTargetRow = onPaste({
-      sourceRow: copiedCell.row,
-      sourceColumnKey: copiedCell.columnKey,
-      targetRow,
-      targetColumnKey: targetColumn.key
-    });
+    const cellEvent = createCellEvent(event);
+    const updatedTargetRow = onPaste(
+      {
+        sourceRow: copiedCell.row,
+        sourceColumnKey: copiedCell.columnKey,
+        targetRow,
+        targetColumnKey: targetColumn.key
+      },
+      cellEvent
+    );
+    if (cellEvent.isGridDefaultPrevented()) return;
 
     updateRow(targetColumn, rowIdx, updatedTargetRow);
```

Comment:
if developers use `event.preventGridDefault()` in `onPaste` handler, the default behavior to update the row would be prevented.